Vault Init script

v-starodubov opened the pull request:
This is small rework of #3 request, related to #2 

Changes:
- Lightweight and secure Docker image that contains all scripts required.
- Init scripts reworked to make API requests directly to k8s API server and Vault API without requiring installed kubectl (and corresponding RBAC permissions in cluster)
- Removed unnecessary dynamic RBAC manifests creation. We don't need to modify them dynamically.
- Deleted `external-secrets` dependency, because right now it don't have required namespace label for umbrella chart deployment ([Changes requested](https://github.com/external-secrets/external-secrets/pull/3082))
- Removed ESO manifests and created templates/tests cases for `helm test` runs (requires installed ESO)
- Added helm repository, hosted here in `helm-repo` branch.
- GitHub Actions: build and scan Docker image with Trivy (triggered on tag with proper semver formating), helm packaging and deploying on changes in `vault/Chart.yaml`.
- Updated README.md with proper commands.

h0neybadger3 committed:
Init chart

v-starodubov committed:
Rework Dockerfile to single image and vault init rework from kubectl to curl API requests

v-starodubov committed:
Fixes to vault init script

v-starodubov committed:
Rework of init ESO script and small fixes

v-starodubov committed:
Update chart dependencies

v-starodubov committed:
Create test hook for external-secrets testing

v-starodubov committed:
Refactor main deployment chart

v-starodubov committed:
Remove external-secrets dependency and ad ca certs to request

v-starodubov committed:
Add github action to deploy charts and image, remove Chart lock

v-starodubov committed:
Change image scanner to Trivy

v-starodubov committed:
Test build and scan image using tarball

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Update actions trigger paths

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov commented:
Currently this branch is not fully tested and can contain some errors during run.
I'll update this PR with `ready to review` when everything is done.

v-starodubov committed:
Fix helm lint errors, wrong name links and cacert use in curl command

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Fix errors in chart manifests and init scripts

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Fix typo in job manifest

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov commented:
@jradikk ready to review.

jradikk commented on .github/workflows/docker-build-scan.yaml:
```diff
@@ -0,0 +1,111 @@
+          ignore-unfixed: true
+          severity: 'CRITICAL,HIGH'
+          exit-code: '1'
+          format: 'table'
```
I think Github supports using nice reports for security findings. Although, it might require a subscription. Can you check that please? If it's not an option, let's think about maybe using badges in README or anything else that comes with a nicer UI than a log

v-starodubov replied:
There two possible ways:
- [SBOM report](https://github.com/marketplace/actions/aqua-security-trivy#using-trivy-to-generate-sbom) to Github Dependency Graph.
- [SARIF report](https://github.com/marketplace/actions/aqua-security-trivy#using-trivy-with-github-code-scanning) to Github Code Scanning.

I don't know which is better, so I'll test both.

v-starodubov replied:
Choosed SARIF code scanning, because SBOM not working with non-default branches and tags. Example of medium severity report: https://github.com/Alpacked/security-hardening-helm/security/code-scanning/1

Another question about severity of reports, we can have all types in Security tab and GitHub will create alerts only for high and higher thresholds.
Or left or mechanism as is, create security findings only for HIGH and CRITICAL vulnerabilities?
<img width="967" alt="Screenshot 2024-02-14 at 15 24 37" src="https://github.com/Alpacked/security-hardening-helm/assets/3694552/a276aea9-7686-4e92-b5b4-ac0cdfb4dbef">
<img width="967" alt="Screenshot 2024-02-14 at 15 19 04" src="https://github.com/Alpacked/security-hardening-helm/assets/3694552/87c07e91-f091-47be-a118-524e10469e9a">

jradikk replied:
Let's capture all of them, but fail the build and alert for high and critical only

v-starodubov replied:
If we want use SARIF/Code scanning reports, we can't fail our pipeline during scan action because it will prevent results from uploading. I'm thinking on two options:
1. (Lame) Make second scan right before pushing to Docker Hub.
2. Build and scan at (master) branch push (restrict to paths: scripts/*.sh and Dockerfile), build and push at tag push (will be restricted by ruleset if there high or higher errors not resolved).

What do you think?

jradikk replied:
Is there an option to create a some sort of an artifact or a variable that would contain a boolean value, based on which a push job is triggered?

v-starodubov replied:
Found a workaround for how to run the upload step even if 'scan' found some vulnerabilities, but if SARIF isn't limited by our rule (we can upload a full file with 'limit-severities-for-sarif' parameter) it will always fail if some vulnerabilities are found.

So in that case we need to limit SARIF output to our HIGH and CRITICAL, until trivy-action change something.

v-starodubov replied:
Created issue in trivy-action repository: https://github.com/aquasecurity/trivy-action/issues/309

For now, security findings will be limited by high and critical vulnerabilities.

jradikk commented on README.md:
```diff
@@ -0,0 +1,48 @@
+
+We use additional scripts to initialize and unseal Vault pods. It can also add permissions for External Secrets operator that uses KV2 store.
+
+As result of executing this chart you should have created 'vault-init-secrets' with corresponding values that used for initialization. Note that this secret isn't going deleted or re-written after deleting release, so you must manually ensure it's rotation and safety.
```
Run this part by grammarly or similar. There seems to be few mistakes

jradikk commented on scripts/vault-init-eso.sh:
```diff
@@ -0,0 +1,67 @@
+  curl -X POST \
+      --header "X-Vault-Token: ${VAULT_ROOT_TOKEN}" \
+      --data "{\"type\": \"kv\", \"options\": {\"version\": \"2\"}}" \
+      "${VAULT_ADDRESS}/v1/sys/mounts/kv"
```
How about moving this curl command to a separate function or a variable? We copypaste a part of it multiple times changing only payload and URL

jradikk commented on scripts/vault-init-eso.sh:
```diff
@@ -0,0 +1,67 @@
+      "${VAULT_ADDRESS}/v1/auth/kubernetes/role/eso-reader"
+}
+
+get_vault_root_token
```
What do you think about moving echo from functions to this place of their execution? I feel like it should ease the process of reading the script

jradikk commented on README.md:
```diff
@@ -0,0 +1,48 @@
+
+As result of executing this chart you should have created 'vault-init-secrets' with corresponding values that used for initialization. Note that this secret isn't going deleted or re-written after deleting release, so you must manually ensure it's rotation and safety.
+
+Please note that for External Secrets initialization with Vault the ESO should be already deployed. Helm test command will create temporary secret to test connection between ESO and Vault.
```
Please extend the README explaining what exactly does this chart do

jradikk commented on vault/templates/job-eso.yaml:
```diff
@@ -0,0 +1,42 @@
+{{- if .Values.esoInit.enabled -}}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ .Chart.Name }}-eso-init
```
what's the point of having multiple templates for jobs? I don't see much difference between them, so I think we may combine them into a single job template and instead of using `{{- if .Values.esoInit.enabled -}}` we can just iterate the jobs object

jradikk commented on vault/templates/role-vault.yaml:
```diff
@@ -0,0 +1,20 @@
+  - apiGroups: [""]
+    resources: ["pods"]
+    verbs: ["list"]
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["create"]
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["update"]
```
you can combine verbs into a single array

v-starodubov replied:
I'm using separate array for 'update' because it's limited to corresponding resource 'vault-init-secrets'.
```yaml
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["create", "update"]
    resourceNames: ["vault-init-secrets"]
```
If i use combined array for this action we'll get error on creating secret:
```
"secrets is forbidden: User \"system:serviceaccount:vault-system:auth-vault-init\" cannot create resource \"secrets\" in API group \"\" in the namespace \"vault-system\""
```

jradikk commented on vault/templates/rolebinding-eso.yaml:
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.esoInit.enabled }}
```
The same question as before about jobs. I think we need a single template for rolebinding

jradikk commented on vault/values.yaml:
```diff
@@ -0,0 +1,31 @@
+    ha:
+      enabled: false
+      raft:
+        enabled: false
```
It is generally not recommended to use nesting that much in helm unless there's a valid reason for it. I recommend 
1) flattening the structure to something like `vaultInjector: true` instead of vault.injector.enabled
2) hide default settings from values.yaml and instead document them in README

v-starodubov replied:
These values are belong to dependended 'vault' chart, so i don't know if we can't change them 🤔

jradikk commented on vault/values.yaml:
```diff
@@ -0,0 +1,31 @@
+        enabled: false
+
+vaultInit:
+  serviceAccount:
+    name: "auth-vault-init"
```
I think that has to be a default value, hidden from values.yaml. Because I don't think there'll be much need for renaming the service account

v-starodubov committed:
Flattering the chart
- Removing vaultInit.serviceAccount
- Moving image values to values root
- Iterating for enabled inits for manifests: jobs, roles, rolebindings

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Changing scan step to upload SBOM to Github Dependency Graph

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Making DRY script changes

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Reformat the intro in README.md

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Use SARIF file for security findings

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Update README.md with helm-docs

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Update chart version and fix typo

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Fix typo substitution and add some info echoes to test script

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Remove cves limiting in SAFIR reports

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>

v-starodubov committed:
Update scan action to always upload SARIF report

Signed-off-by: Volodymyr Starodubov <volodymyr.starodubov@alpacked.io>
